warehouse/tracker.go

```diff
@@ -0,0 +1,183 @@
+			rruntime.GoForWarehouse(func() {
+				defer wg.Done()
+				ch <- warehouse
+			})
+		}
+		wh.configSubscriberLock.RUnlock()
+
+		rruntime.Go(func() {
+			wg.Wait()
+			close(ch)
+		})
+
+		for w := range ch {
+			if err := wh.Track(ctx, &w, config.Default); err != nil {
+				return fmt.Errorf("cron tracker failed for source: %s, destination: %s with error: %w", w.Source.ID, w.Destination.ID, err)
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			wh.Logger.Infof("context is cancelled, stopped running tracking")
+			return nil
+		case <-time.After(uploadStatusTrackFrequency):
+		}
+	}
+}
+
+func (wh *HandleT) Track(ctx context.Context, warehouse *warehouseutils.Warehouse, config *config.Config) error {
+	var (
+		query         string
+		queryArgs     []interface{}
+		createdAt     sql.NullTime
+		exists        bool
+		uploaded      int
+		syncFrequency = "1440"
+		Now           = timeutil.Now
+		NowSQL        = "NOW()"
+		timeWindow    = config.GetInt("Warehouse.uploadBufferTimeInMin", 180)
+		source        = warehouse.Source
+		destination   = warehouse.Destination
+	)
+
+	if wh.NowSQL != "" {
+		NowSQL = wh.NowSQL
+	}
+	if wh.Now != nil {
+		Now = wh.Now
+	}
+
+	if !source.Enabled || !destination.Enabled {
+		return nil
+	}
+
+	excludeWindow := warehouseutils.GetConfigValueAsMap(warehouseutils.ExcludeWindow, warehouse.Destination.Config)
+	excludeWindowStartTime, excludeWindowEndTime := GetExcludeWindowStartEndTimes(excludeWindow)
+	if CheckCurrentTimeExistsInExcludeWindow(Now(), excludeWindowStartTime, excludeWindowEndTime) {
+		return nil
+	}
+
+	if sf := warehouseutils.GetConfigValue(warehouseutils.SyncFrequency, *warehouse); sf != "" {
+		syncFrequency = sf
+	}
+	if value, err := strconv.Atoi(syncFrequency); err == nil {
+		timeWindow += value
+	}
+
+	query = fmt.Sprintf(`
+				SELECT
+				  created_at
+				FROM
+				  %[1]s
+				WHERE
+				  source_id = '%[2]s' AND
+				  destination_id = '%[3]s' AND
+				  created_at > %[6]s - interval '%[4]d MIN' AND
+				  created_at < %[6]s - interval '%[5]d MIN'
+				ORDER BY
+				  created_at DESC
+				LIMIT
+				  1;
+				`,
+		warehouseutils.WarehouseStagingFilesTable,
+		source.ID,
+		destination.ID,
+		2*timeWindow,
+		timeWindow,
+		NowSQL,
+	)
+
+	err := wh.dbHandle.QueryRowContext(ctx, query).Scan(&createdAt)
+	if err == sql.ErrNoRows {
+		return nil
+	}
+	if err != nil && err != sql.ErrNoRows {
+		return fmt.Errorf("fetching last upload time for source: %s and destination: %s: %w", source.ID, destination.ID, err)
+	}
+
+	if !createdAt.Valid {
+		return fmt.Errorf("invalid last upload time for source: %s and destination: %s", source.ID, destination.ID)
+	}
+
+	query = fmt.Sprintf(`
+				SELECT
+				  EXISTS (
+					SELECT
+					  1
+					FROM
+					  %s
+					WHERE
+					  source_id = $1 AND
+					  destination_id = $2 AND
+					  (
+						status = $3
+						OR status = $4
+						OR status LIKE $5
+					  ) AND
+					  updated_at > $6
+				  );
+	`,
+		warehouseutils.WarehouseUploadsTable,
```

Comment:
Personal preference, assuming `warehouseutils.WarehouseUploadsTable` is a constant
```suggestion
	query = `
				SELECT
				  EXISTS (
					SELECT
					  1
					FROM
					  ` + warehouseutils.WarehouseUploadsTable + `
					WHERE
					  source_id = $1 AND
					  destination_id = $2 AND
					  (
						status = $3
						OR status = $4
						OR status LIKE $5
					  ) AND
					  updated_at > $6
				  );
	`
```